blueocean-pipeline-api-impl/src/test/java/io/jenkins/blueocean/rest/impl/pipeline/GraphBuilderTest.java

```diff
@@ -409,6 +412,53 @@ public void unstableSmokes() throws Exception {
         assertEquals("Unexpected stages in graph", 4, nodes.size());
     }
 
+    /**
+     * Builds a basic Pipeline where the overall graph structure (in terms of stages) is the same from one build
+     * to the next, but there are differences in the number of FlowNodes (and thus different node IDs for
+     * corresponding nodes like the starts of stages) in each build.
+     *
+     * If this test breaks, it probably means that something is mixing up node IDs across different builds when it
+     * should not be doing so.
+     */
+    @Test
+    public void unionDifferentNodeIdsSameStructure() throws Exception {
+        WorkflowJob p = createJob("unionDifferentNodeIdsSameStructure", "unionDifferentNodeIdsSameStructure.jenkinsfile");
+        // Run the first build, it should complete successfully. We don't care about its structure.   
+        WorkflowRun b1 = p.scheduleBuild2(0).waitForStart();
+        SemaphoreStep.waitForStart("second/1", b1);
+        SemaphoreStep.success("second/1", null);
+        SemaphoreStep.waitForStart("third/1", b1);
+        SemaphoreStep.success("third/1", null);
+        j.waitForCompletion(b1);
+        j.assertBuildStatus(Result.SUCCESS, b1);
+        // Run the second build. Its graph as computed by PipelineNodeGraphVisitor will be created by combining the current build with the previous build.
+        WorkflowRun b2 = p.scheduleBuild2(0).waitForStart();
+        SemaphoreStep.waitForStart("second/2", b2);
+        {
+            List<FlowNodeWrapper> nodes = unionPipelineNodes(b1, b2);
+            assertStageAndEdges(nodes, "first", BlueRun.BlueRunState.FINISHED, BlueRun.BlueRunResult.SUCCESS, "second");
+            assertStageAndEdges(nodes, "second", BlueRun.BlueRunState.RUNNING, BlueRun.BlueRunResult.UNKNOWN, "third");
```

Comment:
If you revert the changes in `src/main`, all assertions expecting `BlueRunState.RUNNING` and `BlueRunResult.UNKNOWN` fail because they are `null` instead.